### DISCUSSION TITLE
I don't understand why this machine is stuck in it's initial state

### DISCUSSION THREAD

**Discussion by karfau** [statelyai/xstate#1565] [Q&A] [answered]
Hey there,

I am trying to build a state machine that models (in the end async) form field validation.
I have tried tons of different things but I'm quite sure there is no general issue with it,

but somehow the machine never leaves it's initial state (even when the guard is visualized as "green"):

https://gist.github.com/karfau/7711b52a5cb2e43e59d24b00e3c60b53

https://xstate.js.org/viz/?gist=7711b52a5cb2e43e59d24b00e3c60b53

![image](https://user-images.githubusercontent.com/135657/96349783-28d1e700-10b2-11eb-87c2-133d56fd9c62.png)

I also know that it should work/be enough to add a top level `on: {CHANGE}` but since it didn't work as expected I copied it into every state.

Any help is much appriciated.

**Comment by rjdestigter** [comment, marked as answer]
```
const CHANGE = [
        {
          actions: ['log', 'storeCurrent']
        },
        {
          cond: 'isChanged',
          target: 'validating'
        },
        {target: 'unchanged'}
      ]
  ````
Because it will never run the the second action here since the first action has no condition and will be picked.

```
const CHANGE = [{
         actions: ['log', 'storeCurrent']
         cond: 'isChanged',
         target: 'validating'
        },
        {target: 'unchanged'}
      ]
  ````

**Reply by karfau** [reply]
I wanted to ensure that the actions are always executed (and before the guard is called).

Is that he case in your suggested solution?

**Reply by davidkpiano** [reply]
> I wanted to ensure that the actions are always executed (and before the guard is called).

There was a previous thread about this, but in reality, this is either two states (one in which the actions are executed and you `always` go to the second state, where the guard is applied), or copying the action for each branch.

**Reply by karfau** [reply]
It makes sense in that guards need to also be able to prevent actions.
Thx for the fast responses and the helpful explanation.

Update: For people reading that later I applied the suggested fix I found  most applicable (and some other fixes) in a new gist:
https://gist.github.com/karfau/760d85cb7ed38017a31f958fef883c7e

https://xstate.js.org/viz/?gist=760d85cb7ed38017a31f958fef883c7e

![image](https://user-images.githubusercontent.com/135657/96352017-60e02680-10c0-11eb-8948-8d1900fa291b.png)
